feat: initial portal docs + minor cleanups (#1469)
AztecProtocol/aztec-packages: docs/docs/dev_docs/contracts/portals/main.md

  fn message_portal(&mut self, content: Field)
```

When sending a message from L2 to L1 we don't need to pass recipient, deadline, secret nor fees. Recipient is populated with the attached portal and the remaining values are not needed as the message is inserted into the outbox at the same time as it was included in a block (for the inbox it could be inserted and then only included in rollup block later). 

:::danger
Need to have access control on the L1 portal contract to not consume messages that were sent from the wrong L2 contract.
:::

As earlier, we can use a token bridge as an example. In this case, we are burning tokens on L2 and sending a message to the portal to free them on L1.

```rust title="NonNativeTokenContract.nr"
// burns token and sends a message to the portal
fn withdraw(
    amount: Field,
    sender: Field,
    recipient: Field,
    callerOnL1: Field,
) -> distinct pub abi::PrivateCircuitPublicInputs {
    ...
    let sender_balance = storage.balances.at(sender);
    spend_notes(&mut context, sender_balance, amount, sender);

    let content = get_withdraw_content_hash(amount, recipient, callerOnL1);
    context.message_portal(content);
    ...
}
```

> When the transaction is included in a rollup block it will be inserted into the `Outbox`, where the recipient portal can consume it from. When consuming, the `msg.sender` must match the `recipient` meaning that only portal can actually consume the message.

Reviewer: ```suggestion
When the transaction is included in a rollup block the message will be inserted into the `Outbox`, where the recipient portal can consume it from. When consuming, the `msg.sender` must match the `recipient` meaning that only portal can actually consume the message.
```

```solidity title="TokenPortal.sol"
function withdraw(uint256 _amount, address _recipient, bool _withCaller)
    external
    returns (bytes32)
  {
    // Create the message structure
    DataStructures.L2ToL1Msg memory message = DataStructures.L2ToL1Msg({
      sender: DataStructures.L2Actor(l2TokenAddress, version),
      recipient: DataStructures.L1Actor(address(this), block.chainid),
      content: Hash.sha256ToField(
        abi.encodeWithSignature(
          "withdraw(uint256,address,address)",
          _amount,
          _recipient,
          _withCaller ? msg.sender : address(0)
        )
        )
    });

    // Consume the message
    bytes32 entryKey = registry.getOutbox().consume(message);

    // Transfer the tokens to the user
    underlying.transfer(_recipient, _amount);

    return entryKey;
  }
```

## How to deploy a contract with a portal
- Deploy L1 
- Deploy l2
- Initialize l1 with l2 address.

## Standards

### Structure of messages

> The application developer should consider creating messages that follow a structure like function calls, e.g., using a function signature and arguments. This will make it easier to ensure that they are not making messages that could be misinterpreted by the recipient. 

Reviewer: ```suggestion
The application developer should consider creating messages that follow a function call structure e.g., using a function signature and arguments. This will make it easier to prevent producing messages that could be misinterpreted by the recipient. 
```

An example of a bad format would be using `amount, token_address, recipient_address` as the message for a withdraw function and `amount, token_address, on_behalf_of_address` for a deposit function. Any deposit could then also be mapped to a withdraw or vice versa. 

```solidity 
// Don't to this!
bytes memory message = abi.encode(
  _amount, 
  _token, 
  _to
);

// Do this!
bytes memory message abi.encodeWithSignature(
  "withdraw(uint256,address,address)", 
  _amount, 
  _token, 
  _to
);

> ```

Reviewer 2: <3

### Error Handling

> Handling error when moving cross chain can quickly get tricky. Since the L1 and L2 calls are practically async and independent of each other, the L1 part of a deposit might execute just fine, with the L2 part failing. If not handling this well, the funds might be lost forever! The contract builder should therefore consider in what ways his application can fail cross chain, and handle those cases explicitly.

Reviewer: ```suggestion
Handling error when moving cross chain can quickly get tricky. Since the L1 and L2 calls are practically async and independent of each other, the L1 part of a deposit might execute just fine, with the L2 part failing. If this is not handled well, the funds may be lost forever! The contract builder should therefore consider ways their application can fail cross chain, and handle all cases explicitly.
```

> First, entries in the outboxes **SHOULD** only be consumed if the execution is successful. For an L2 -> L1 call, the L1 execution can revert the transaction completely if anything fails. As the tx can be atomic, the failure also reverts the consumption of the entry.

Reviewer: ```suggestion
First, entries in the outboxes **SHOULD** only be consumed if the execution is successful. For an L2 -> L1 call, the L1 execution can revert the transaction completely if anything fails. As the tx is atomic, the failure also reverts consumption.
```

If it is possible to enter a state where the second part of the execution fails forever, the application builder should consider including additional failure mechanisms (for token withdraws this could be depositing them again etc).

Generally it is good practice to keep cross-chain calls simple to avoid too many edge cases and state reversions.

:::info
Error handling for cross chain messages is handled by the application contract and not the protocol. The protocol only delivers the messages, it does not ensure that they are executed successfully.
:::

### Cancellations

> A special type of errors is underpriced transactions, here meaning that message is inserted on L1, but the fee available to the sequencer is too low to be included in a rollup block.

Reviewer: ```suggestion
A special type of error is an underpriced transaction, in this context it means that a message is inserted on L1, but the attached fee is too low to be included in a rollup block.
```

> For the case of token bridges, this could lead to funds being locked in the bridge forever, as the funds are locked but the message never arrives on L2 to mint the tokens. To address this, the `Inbox` supports cancelling messages after a deadline. However, this must be called by the portal itself, as it will need to "undo" the state changes is made (for example by sending the tokens back to the user).

Reviewer: ```suggestion
For the case of token bridges, this could lead to funds being locked in the bridge forever, as funds are locked but the message never arrives on L2 to mint the tokens. To address this, the `Inbox` supports cancelling messages after a deadline. However, this must be called by the portal itself, as it will need to "undo" the state changes is made (for example by sending the tokens back to the user).
```

    DataStructures.L1ToL2Msg memory _message, 
    address _feeCollector
) external returns (bytes32 entryKey);
```

Building on our token example from earlier, this can be called like:

```solidity title="TokenPortal.sol"
function cancelL1ToAztecMessage(
    bytes32 _to,
    uint256 _amount,
    uint32 _deadline,
    bytes32 _secretHash,
    uint64 _fee
  ) external returns (bytes32) {
    IInbox inbox = registry.getInbox();
    DataStructures.L1Actor memory l1Actor = DataStructures.L1Actor(address(this), block.chainid);
    DataStructures.L2Actor memory l2Actor = DataStructures.L2Actor(l2TokenAddress, 1);
    DataStructures.L1ToL2Msg memory message = DataStructures.L1ToL2Msg({
      sender: l1Actor,
      recipient: l2Actor,
      content: Hash.sha256ToField(
        abi.encodeWithSignature("mint(uint256,bytes32,address)", _amount, _to, msg.sender)
        ),
      secretHash: _secretHash,
      deadline: _deadline,
      fee: _fee
    });
    bytes32 entryKey = inbox.cancelL2Message(message, address(this));
    // Ensures that `msg.sender == canceller` by using `msg.sender` in the hash computation.
    underlying.transfer(msg.sender, _amount);
    return entryKey;
  }
```

> The example above ensure that the user can cancel his message if it is underpriced.

Reviewer: ```suggestion
The example above ensure that the user can cancel their message if it is underpriced.
```

### Designated caller

> A designated caller is the ability to specify who should be able to call a function that consumes a message. This is useful for ordering of batched messages, as we will see in a second.  

Reviewer: ```suggestion
Designating a caller grants the ability to specify who should be able to call a function that consumes a message. This is useful for ordering of batched messages, as we will see shortly.  
```

> When doing multiple cross-chain calls as one action it is important to consider the order of the calls. Say for example, that you want to do a uniswap trade on L1 because you are a whale and slippage on L2 is too damn high. 

Reviewer: ```suggestion
When performing multiple cross-chain calls in one action it is important to consider the order of the calls. Say for example, that you want to perform a uniswap trade on L1 because you are a whale and slippage on L2 is too damn high. 
```

You would practically, withdraw funds from the rollup, swap them on L1, and then deposit the swapped funds back into the rollup. This is a fairly simple process, but it requires that the calls are done in the correct order. For one, if the swap is called before the funds are withdrawn, the swap will fail. And if the deposit is called before the swap, the funds might get lost! 

> As the message boxes only will allow the recipient portal to consume the message, we can use this to our advantage to ensure that the calls are done in the correct order. Say that we include a designated "caller" in the messages, and that the portal contract checks that the caller matches the designated caller or designated is address(0) (anyone can call). When the message are to be consumed on L1, it can compute the message as seen below:

Reviewer: ```suggestion
As message boxes only will allow the recipient portal to consume the message, we can use this to our advantage to ensure that the calls are done in the correct order. Say that we include a designated "caller" in the messages, and that the portal contract checks that the caller matches the designated caller or designated is address(0) (anyone can call). When the message are to be consumed on L1, it can compute the message as seen below:
```
